notebooks/semi_supervised_hmm.ipynb

```diff
@@ -0,0 +1,503 @@
+    "emission_prob = dist.dirichlet.rvs(emission_prior, num_categories, random_state=rng_emission)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# simulate data\n",
+    "start_prob = np.repeat(1. / num_categories, num_categories)\n",
+    "categories, words = [], []\n",
+    "for t in range(num_supervised_data + num_unsupervised_data):\n",
+    "    rng, rng_transition, rng_emission = random.split(rng, 3)\n",
+    "    if t == 0 or t == num_supervised_data:\n",
+    "        category = dist.categorical.rvs(start_prob, random_state=rng_transition)\n",
+    "    else:\n",
+    "        category = dist.categorical.rvs(transition_prob[category], random_state=rng_transition)\n",
+    "    word = dist.categorical.rvs(emission_prob[category], random_state=rng_emission)\n",
+    "    categories.append(category)\n",
+    "    words.append(word)\n",
+    "categories, words = np.stack(categories), np.stack(words)\n",
+    "\n",
+    "# split into supervised data and unsupervised data\n",
+    "supervised_categories = categories[:num_supervised_data]\n",
+    "supervised_words = categories[:num_supervised_data]\n",
+    "unsupervised_words = categories[num_supervised_data:]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "num_warmup_steps = num_samples = 150"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### NumPyro"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def forward_log_prob(prev_log_prob, curr_word, transition_log_prob, emission_log_prob):\n",
+    "    log_prob = np.expand_dims(prev_log_prob, axis=1) + transition_log_prob\n",
+    "    log_prob = log_prob + emission_log_prob[:, curr_word]\n",
+    "    return logsumexp(log_prob, axis=0)\n",
+    "\n",
+    "def semi_supervised_hmm(supervised_categories, supervised_words, unsupervised_words):\n",
+    "    transition_prob = sample(\"transition_prob\", dist.dirichlet(\n",
+    "        np.broadcast_to(transition_prior, (num_categories, num_categories))))\n",
+    "    emission_prob = sample(\"emission_prob\", dist.dirichlet(\n",
+    "        np.broadcast_to(emission_prior, (num_categories, num_words))))\n",
+    "\n",
+    "    sample(\"supervised_categories\", dist.categorical(transition_prob[supervised_categories[:-1]]),\n",
+    "           obs=supervised_categories[1:])\n",
+    "    sample(\"supervised_words\", dist.categorical(emission_prob[supervised_categories]),\n",
+    "           obs=supervised_words)\n",
+    "\n",
+    "    transition_log_prob = np.log(transition_prob)\n",
+    "    emission_log_prob = np.log(emission_prob)\n",
+    "    log_prob = emission_log_prob[:, unsupervised_words[0]]\n",
+    "    # TODO: to be replaced by lax.scan when its jvp rule is implemented\n",
```

Comment:
Unfortunately, I think this will be a blocker to getting reasonable performance on these kind of models.